### PR TITLE
*: upgrade to go 1.20.8

### PR DESCRIPTION
**Pull request by healthy-pod**
* [x] Adjust the Pebble tests to run in new version.
* [x] Adjust version in Docker image ([source](./builder/Dockerfile)).
* [x] Adjust version in the TeamCity agent image ([setup script](./packer/teamcity-agent.sh))
* [x] Rebuild and push the Docker image (following [Basic Process](#basic-process))
* [x] Update `build/teamcity/internal/release/build-and-publish-patched-go/impl.sh` with the new version and adjust SHA256 sums as necessary.
* [x] Adjust `GO_VERSION` and `GO_FIPS_COMMIT` for the FIPS Go toolchain ([source](./teamcity/internal/release/build-and-publish-patched-go/impl-fips.sh)).
* [x] Run the `Internal / Cockroach / Build / Toolchains / Publish Patched Go for Mac` build configuration in TeamCity with your latest version of the script above. Note the job depends on another job `Build and Publish Patched Go`. That job prints out the SHA256 of all tarballs, which you will need to copy-paste into `WORKSPACE` (see below). `Publish Patched Go for Mac` is an extra step that publishes the *signed* `go` binaries for macOS. That job also prints out the SHA256 of the Mac tarballs in particular.
* [x] Adjust `--@io_bazel_rules_go//go/toolchain:sdk_version` in [.bazelrc](../.bazelrc).
* [x] Bump the version in `WORKSPACE` under `go_download_sdk`. You may need to bump [rules_go](https://github.com/bazelbuild/rules_go/releases). Also edit the filenames listed in `sdks` and update all the hashes to match what you built in the step above.
* [x] Bump the version in `WORKSPACE` under `go_download_sdk` for the FIPS version of Go (`go_sdk_fips`).
* [x] Run `./dev generate bazel` to refresh `distdir_files.bzl`, then `bazel fetch @distdir//:archives` to ensure you've updated all hashes to the correct value.
* [x] Bump the version in `builder.sh` accordingly ([source](./builder.sh#L6)).
* [x] Bump the version in `go-version-check.sh` ([source](./go-version-check.sh)), unless bumping to a new patch release.
* [x] Bump the go version in `go.mod`.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh)).
* [x] Replace other mentions of the older version of go (grep for `golang:<old_version>` and `go<old_version>`).
* [x] Update the `builder.dockerImage` parameter in the TeamCity [`Cockroach`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Cockroach&tab=projectParams) and [`Internal`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Internal&tab=projectParams) projects.
* [x] Ask the Developer Infrastructure team to deploy new TeamCity agent images according to [packer/README.md](./packer/README.md)

Release note (build change): upgrade to go 1.20.8
Epic: none

### CODE REVIEW

**Review by rharding6373 — COMMENTED** [review]
:lgtm: for sql queries

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109773#-:-NdBMrX9-5H83FHRQBdp:bcq9lgu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lidorcarmel, @renatolabs, and @smg260)

File: build/teamcity/internal/release/build-and-publish-patched-go/diff.patch [hunk 1]
```diff
@@ -20,26 +18,40 @@ index 3991a48b10..4a50ae650d 100644
 +		// state.
 +		gp.runningnanos += now - gp.lastsched
```
Comment (srosenberg) [inline build/teamcity/internal/release/build-and-publish-patched-go/diff.patch:19]: I noticed that a histogram tracking scheduling latencies has been added [1]. Technically, the sum of all latencies is now derivable via `runtime.metrics`. Do we still need to maintain this patch?

[1] https://github.com/golang/go/commit/bedfeed54a7a80123c07f83c325a0bcfe5c43398

**Review by srosenberg — APPROVED** [review]
Thanks!

File: build/teamcity/internal/release/build-and-publish-patched-go/diff.patch (hunk 1 above)
Comment (healthy-pod) [reply build/teamcity/internal/release/build-and-publish-patched-go/diff.patch:19]: cc @irfansharif

Comment (irfansharif) [reply build/teamcity/internal/release/build-and-publish-patched-go/diff.patch:19]: The patch doesn’t measure scheduling latencies, it measures how much running time each goroutine gets on-CPU. Yes, we still need it.

Comment (srosenberg) [reply build/teamcity/internal/release/build-and-publish-patched-go/diff.patch:19]: Yep, I see that now; thanks.

```
+       if newval == _Grunning {
+               // We're transitioning into the running state, record the timestamp for
+               // subsequent use.
+               gp.lastsched = now
+       }
+
        if oldval == _Grunning {
+               // We're transitioning out of running, record how long we were in the
+               // state.
+               gp.runningnanos += now - gp.lastsched
```

**Review by rail — APPROVED** [review]

**Review by nvb — APPROVED** [review]
:lgtm: for kv. Thanks for pushing on this @healthy-pod.

Here are a pair of issues for improvements that we can make once this upgrade lands:
- https://github.com/cockroachdb/cockroach/issues/110262
- https://github.com/cockroachdb/cockroach/issues/110261

Also, what are our plans with Go 1.21? It's been out for exactly a month and we have about another month to stability for the v23.2 release. Should we immediately follow up this upgrade with the next one?

> Reviewed 19 of 29 files at r1, 7 of 9 files at r2, 21 of 21 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109773#-:-NdpVvxZ-WA65tgIkpcp:b-osyo42)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @healthy-pod, @irfansharif, @lidorcarmel, @renatolabs, and @smg260)

___
*[`pkg/util/schedulerlatency/testdata/histogram_buckets` line 105 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/109773#-NdpVGNU02f1wxJkiDyP:-NdpVGNU02f1wxJkiDyQ:bckqiy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/78fee9932fa25658a66043cd5c55a8908585e589/pkg/util/schedulerlatency/testdata/histogram_buckets#L105)):*
> ```
> ----
> bucket[  0] width=0s                 boundary=[-Inf, 0s)
> bucket[  1] width=64ns               boundary=[0s, 64ns)
> ```

Do we know which change to the go stdlib caused this reduction in histogram buckets in the `runtime/metrics` package? @irfansharif are you aware of this change?

**Comment by healthy-pod** [comment]
> Do we know which change to the go stdlib caused this reduction in histogram buckets in the runtime/metrics package?

related comment: https://github.com/cockroachdb/cockroach/pull/97260#pullrequestreview-1314175304

**Comment by healthy-pod** [comment]
> Also, what are our plans with Go 1.21? It's been out for exactly a month and we have about another month to stability for the v23.2 release. Should we immediately follow up this upgrade with the next one?

@nvanbenschoten We were planning to upgrade to Go 1.21 directly but [Go FIPS ](https://github.com/golang-fips/go) 1.21 is not out yet so upgrading non-fips to 1.21 would leave us with two different major Go versions on the same branch which can cause numerous issues. We decided to keep non-fips Go major version and the fips one in sync on the same branch at all times. Once Go FIPS 1.21 is out we can upgrade to Go 1.21 immediately.

**Comment by nvanbenschoten** [comment]
> Once Go FIPS 1.21 is out we can upgrade to Go 1.21 immediately.

https://github.com/golang-fips/go/pull/114 merged yesterday. Is that what we were waiting for?

**Comment by healthy-pod** [comment]
> https://github.com/golang-fips/go/pull/114 merged yesterday. Is that what we were waiting for?

Nice, we will still wait until a `go1.21-fips-release` release branch is cut.

**Comment by rafiss** [comment]
This PR has changed the `uuid.FastV4` function to be backed by `crypto/rand` instead of `math/rand`. Was that intentional? In https://github.com/cockroachdb/cockroach/issues/30236#issuecomment-434588162 @nvanbenschoten found that the usage of `crypto/rand` would block on waiting for more entropy, and could affect performance in some workloads.

**Pull request merged**